File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* solve(TreeNode* root){
        if(root == NULL || (root->left == NULL && root->right == NULL)) return root;

        TreeNode* leftTree = solve(root->left);
        TreeNode* rightTree = solve(root->right);
        
        if(leftTree){
            leftTree->right = root->right;
            root->right = root->left;
            root->left = NULL;
        }
        if(rightTree) return rightTree;
        return leftTree;
        // return root;
    }
    void flatten(TreeNode* root) {
        solve(root);
    }
};

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if(root == null) return;
        flatten(root.left);
        flatten(root.right);
        if(root.left != null){
            TreeNode t1 = root.left;
            TreeNode t2 = root.right;
            root.right = t1;
            root.left = null;
            while(root.right!=null) root = root.right;
            root.right = t2;
        }
    }
}